sched/signal/sig_procmask.c

```diff
@@ -187,8 +187,27 @@ int nxsig_procmask(int how, FAR const sigset_t *set, FAR sigset_t *oset)
 
 int sigprocmask(int how, FAR const sigset_t *set, FAR sigset_t *oset)
 {
+  sigset_t nset;
   int ret;
 
+  /* SIGKILL and SIGSTOP should not be added to signal mask */
+
+  if (set != NULL)
+    {
+      nset = *set;
+      if (nxsig_ismember(&nset, SIGKILL))
+        {
+          nxsig_delset(&nset, SIGKILL);
+        }
+
+      if (nxsig_ismember(&nset, SIGSTOP))
+        {
+          nxsig_delset(&nset, SIGSTOP);
+        }
+
+      set = &nset;
+    }
+
   /* Let nxsig_procmask do all of the work */
 
   ret = nxsig_procmask(how, set, oset);
```

Comment:
Note that there are two different instances of sigprocmask:  One common sigprocmask in the task group structure, and one in each TCB.  The latter are used by pthread_kill(), pthread_sigprocmask(), etc.

Shouldn't the logic that removes SIGKILL and SIGSTOP also be in pthread_sigprocmask()?  The effect of sending SIGKILL (or SIGSTOP) via kill(SIGKILL, pid) and pthread_kill(thread, SIGKILL) must be identical.  SIGKILL and SIGSTOP cannot be masked in either and both effect the whole process/task group.

Reply:
NOTE: While a signal handler is running, all signals are blocked (as I recall), including SIGKILL and SIGSTOP) but are restored when the signal handler returns.  This is a case where it is necessary to set even these signals to avoid unhandled re-entrancy.